### PR TITLE
feat(stages): normalized sender recovery commits

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

This PR changes the commit logic for sender recovery stage. It processes transaction batches instead of block batches.

### CODE REVIEW

File: crates/config/src/config.rs [hunk 1]
```diff
@@ -178,13 +178,13 @@ impl From<BodiesConfig> for BodiesDownloaderBuilder {
...
 pub struct SenderRecoveryConfig {
-    /// The maximum number of blocks to process before committing progress to the database.
+    /// The maximum number of transactions to process before committing progress to the database.
     pub commit_threshold: u64,
 }
 
 impl Default for SenderRecoveryConfig {
     fn default() -> Self {
-        Self { commit_threshold: 5_000 }
+        Self { commit_threshold: 50_000 }
```
Comment (rkrasiuk) [inline crates/config/src/config.rs:187]: subject to discussion

File: crates/stages/src/stage.rs [hunk 2]
```diff
@@ -58,6 +69,37 @@ impl ExecInput {
...
+        for entry in body_indices_cursor.walk_range(start_block..=target_block)? {
+            let (block, body) = entry?;
+            last_tx_number = body.last_tx_num();
+            end_block_number = block;
+            let tx_count = (first_tx_number..=last_tx_number).count() as u64; // +1 to include the first tx
+            if tx_count > tx_threshold {
+                break
+            }
+        }
+        Ok((first_tx_number..=last_tx_number, end_block_number, end_block_number >= target_block))
```
Comment (shekhirin) [inline crates/stages/src/stage.rs:101]: nit

```suggestion
        let is_final_range = end_block_number >= target_block;
        Ok((first_tx_number..=last_tx_number, end_block_number, is_final_range))
```

File: crates/stages/src/stage.rs [hunk 3]
```diff
@@ -58,6 +69,37 @@ impl ExecInput {
...
+
+        let first_tx_number = start_block_body.first_tx_num();
+        let mut last_tx_number = start_block_body.last_tx_num();
+        let mut end_block_number = start_block;
+        let mut body_indices_cursor = tx.cursor_read::<tables::BlockBodyIndices>()?;
+        for entry in body_indices_cursor.walk_range(start_block..=target_block)? {
+            let (block, body) = entry?;
+            last_tx_number = body.last_tx_num();
+            end_block_number = block;
+            let tx_count = (first_tx_number..=last_tx_number).count() as u64; // +1 to include the first tx
```
Comment (shekhirin) [inline crates/stages/src/stage.rs:96]: inclusive range already includes the first tx, right? is this an outdated comment?

File: crates/stages/src/stage.rs [hunk 4]
```diff
@@ -58,6 +69,37 @@ impl ExecInput {
         let is_final_range = end == target;
         (start..=end, is_final_range)
     }
+
+    /// Return the next block range determined the number of transactions within it.
+    /// This function walks the the block indices until either the end of the range is reached or
+    /// the number of transactions exceeds the threshold.
+    pub fn next_block_range_with_transaction_threshold<DB: Database>(
```
Comment (shekhirin) [inline crates/stages/src/stage.rs:76]: love the reusability of this function

File: crates/stages/src/stage.rs [hunk 5]
```diff
@@ -58,6 +69,37 @@ impl ExecInput {
...
+        &self,
+        tx: &Transaction<'_, DB>,
+        tx_threshold: u64,
+    ) -> Result<(RangeInclusive<TxNumber>, BlockNumber, bool), StageError> {
```
Comment (shekhirin) [inline crates/stages/src/stage.rs:80]: maybe return the block range, as we do in other functions for consistency?
```suggestion
    ) -> Result<(RangeInclusive<TxNumber>, RangeInclusive<BlockNumber>, bool), StageError> {
```

or change the function comment to reflect that we return only the end of the range

File: crates/stages/src/stage.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/stages/src/stage.rs:96]: right, comment is outdated

**Review by shekhirin — APPROVED** [review]

File: crates/config/src/config.rs (hunk 1 above)
Comment (mattsse) [reply crates/config/src/config.rs:187]: this should be fine 50k, even bettern than 5k

File: crates/stages/src/stages/sender_recovery.rs [hunk 6]
```diff
@@ -178,6 +154,24 @@ impl<DB: Database> Stage<DB> for SenderRecoveryStage {
     }
 }
 
+fn recover_sender(
```
Comment (mattsse) [inline crates/stages/src/stages/sender_recovery.rs:157]: no idea if this is beneficial tbh
```suggestion
#[inline]
fn recover_sender(
```

**Review by mattsse — APPROVED** [review]
makes sense,

haven't reviewed `next_block_range_with_transaction_threshold` closely

File: crates/stages/src/stages/sender_recovery.rs (hunk 6 above)
Comment (rkrasiuk) [reply crates/stages/src/stages/sender_recovery.rs:157]: not sure that inlining is beneficial here

**Review by gakonst — APPROVED** [review]

File: crates/stages/src/stages/sender_recovery.rs [hunk 7]
```diff
@@ -37,7 +37,7 @@ impl SenderRecoveryStage {
 
 impl Default for SenderRecoveryStage {
     fn default() -> Self {
-        Self { commit_threshold: 500_000 }
+        Self { commit_threshold: 50_000 }
```
[the remark this answers is not in the document]
Comment (rkrasiuk) [reply crates/stages/src/stages/sender_recovery.rs:41]: the default value for the stage config was used rather than this, so it doesn't matter

**Pull request merged**